lute/io/src/io.cpp

```diff
@@ -0,0 +1,141 @@
+#include "lute/io.h"
+#include "Luau/Variant.h"
+#include "lute/runtime.h"
+
+#include <uv.h>
+#include <memory>
+
+#include "lua.h"
+#include "lualib.h"
+
+
+namespace io
+{
+
+struct IOHandle
+{
+    Luau::Variant<uv_pipe_t, uv_tty_t> streamVariant;
+    uv_stream_t* stream = nullptr;
+    uv_loop_t* loop = nullptr;
+    ResumeToken resumeToken;
+    lua_State* L = nullptr;
+    std::shared_ptr<IOHandle> self;
+    uv_buf_t *buf;
+
+    ~IOHandle() {
+        if (buf) {
+            free(buf->base);
+            buf = nullptr;
+        }
+    }
+
+    void closeHandles()
+    {
+        auto closeCb = [](uv_handle_t* handle)
+        {
+            IOHandle* ioh = static_cast<IOHandle*>(handle->data);
+            ioh->self.reset();
+        };
+
+        uv_read_stop((uv_stream_t*)stream);
+        uv_close((uv_handle_t*)stream, closeCb);
+    }
+};
+
+static void allocBuffer(uv_handle_t* handle, size_t suggestedSize, uv_buf_t* buf)
+{
+    buf->base = (char*)malloc(suggestedSize);
+    buf->len = buf->base ? suggestedSize : 0;
+    if (!buf->base)
+    {
+        fprintf(stderr, "Process pipe buffer allocation failed!\n");
+    }
+    static_cast<IOHandle*>(handle->data)->buf = buf;
+}
+
+// IOHandle is closed immediately after one read since we don't need a long running stream for this API.
+static void onTtyRead(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf)
+{
+    IOHandle* handle = static_cast<IOHandle*>(stream->data);
+
+    if (nread > 0)
+    {
+        handle->resumeToken->complete(
+            [data = std::string(buf->base, nread)](lua_State* L) -> int
+            {
+                lua_pushlstring(L, data.c_str(), data.size());
+                return 1;
+            }
+        );
+    }
+    else if (nread < 0)
+    {
+        handle->resumeToken->fail("Error reading from stdin");
```

Comment:
I think you can also do `uv_strerror(nread)` here to get the actual error code, and that might also include `UV_EOF`.